File: source/modulo_components/src/LifecycleComponent.cpp
#include "modulo_components/LifecycleComponent.hpp"

#include "modulo_core/exceptions/CoreException.hpp"

using namespace modulo_core::communication;

namespace modulo_components {

LifecycleComponent::LifecycleComponent(const rclcpp::NodeOptions& node_options, const std::string& fallback_name) :
    ComponentInterface<rclcpp_lifecycle::LifecycleNode>(
        node_options, PublisherType::LIFECYCLE_PUBLISHER, fallback_name
    ) {
  this->add_predicate(
      "is_unconfigured", [this] {
        return this->get_current_state().id() == lifecycle_msgs::msg::State::PRIMARY_STATE_UNCONFIGURED;
      }
  );
  this->add_predicate(
      "is_inactive", [this] {
        return this->get_current_state().id() == lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE;
      }
  );
  this->add_predicate(
      "is_active", [this] {
        return this->get_current_state().id() == lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE;
      }
  );
  this->add_predicate(
      "is_finalized", [this] {
        return this->get_current_state().id() == lifecycle_msgs::msg::State::PRIMARY_STATE_FINALIZED;
      }
  );
}

void LifecycleComponent::step() {
  try {
    if (this->get_predicate("is_active")) {
      this->evaluate_periodic_callbacks();
      this->on_step_callback();
      this->publish_outputs();
    }
    this->publish_predicates();
  } catch (const std::exception& ex) {
    RCLCPP_ERROR_STREAM(this->get_logger(), "Failed to execute step function:" << ex.what());
    // TODO handle error in lifecycle component
//    this->raise_error();
  }
}

rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
LifecycleComponent::on_configure(const rclcpp_lifecycle::State& previous_state) {
  RCLCPP_DEBUG(this->get_logger(), "on_configure called from previous state %s", previous_state.label().c_str());
  if (previous_state.id() != lifecycle_msgs::msg::State::PRIMARY_STATE_UNCONFIGURED) {
    RCLCPP_WARN(get_logger(), "Invalid transition 'configure' from state %s.", previous_state.label().c_str());
    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::FAILURE;
  }
  if (!this->handle_configure()) {
    RCLCPP_WARN(get_logger(), "Configuration failed! Reverting to the unconfigured state.");
    // perform cleanup actions to ensure the component is unconfigured
    if (this->handle_cleanup()) {
      return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::FAILURE;
    } else {
      RCLCPP_ERROR(get_logger(),
                   "Could not revert to the unconfigured state! Entering into the error processing transition state.");
      return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::ERROR;
    }
  }
  return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
}

bool LifecycleComponent::handle_configure() {
  bool result = this->on_configure_callback();
  return result && this->configure_outputs();
}

bool LifecycleComponent::on_configure_callback() {
  return true;
}

rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
LifecycleComponent::on_cleanup(const rclcpp_lifecycle::State& previous_state) {
  RCLCPP_DEBUG(this->get_logger(), "on_cleanup called from previous state %s", previous_state.label().c_str());
  if (previous_state.id() != lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE) {
    RCLCPP_WARN(get_logger(), "Invalid transition 'cleanup' from state %s.", previous_state.label().c_str());
    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::FAILURE;
  }
  if (!this->handle_cleanup()) {
    RCLCPP_ERROR(get_logger(), "Cleanup failed! Entering into the error processing transition state.");
    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::ERROR;
  }
  return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
}

bool LifecycleComponent::handle_cleanup() {
  return this->on_cleanup_callback();
}

bool LifecycleComponent::on_cleanup_callback() {
  return true;
}

rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
LifecycleComponent::on_activate(const rclcpp_lifecycle::State& previous_state) {
  RCLCPP_DEBUG(this->get_logger(), "on_activate called from previous state %s", previous_state.label().c_str());
  if (previous_state.id() != lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE) {
    RCLCPP_WARN(get_logger(), "Invalid transition 'activate' from state %s.", previous_state.label().c_str());
    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::FAILURE;
  }
  if (!this->handle_activate()) {
    RCLCPP_WARN(get_logger(), "Activation failed! Reverting to the inactive state.");
    // perform deactivation actions to ensure the component is inactive
    if (this->handle_deactivate()) {
      return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::FAILURE;
    } else {
      RCLCPP_ERROR(get_logger(),
                   "Could not revert to the inactive state! Entering into the error processing transition state.");
      return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::ERROR;
    }
  }
  return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
}

bool LifecycleComponent::handle_activate() {
  bool result = this->activate_outputs();
  return result && this->on_activate_callback();
}

bool LifecycleComponent::on_activate_callback() {
  return true;
}

rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
LifecycleComponent::on_deactivate(const rclcpp_lifecycle::State& previous_state) {
  RCLCPP_DEBUG(this->get_logger(), "on_deactivate called from previous state %s", previous_state.label().c_str());
  if (previous_state.id() != lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE) {
    RCLCPP_WARN(get_logger(), "Invalid transition 'deactivate' from state %s.", previous_state.label().c_str());
    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::FAILURE;
  }
  if (!this->handle_deactivate()) {
    RCLCPP_ERROR(get_logger(), "Deactivation failed! Entering into the error processing transition state.");
    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::ERROR;
  }
  return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
}

bool LifecycleComponent::handle_deactivate() {
  bool result = this->on_deactivate_callback();
  return result && this->deactivate_outputs();
}

bool LifecycleComponent::on_deactivate_callback() {
  return true;
}

rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
LifecycleComponent::on_shutdown(const rclcpp_lifecycle::State& previous_state) {
  RCLCPP_DEBUG(this->get_logger(), "on_shutdown called from previous state %s", previous_state.label().c_str());
  switch (previous_state.id()) {
    case lifecycle_msgs::msg::State::PRIMARY_STATE_FINALIZED:
      return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
    case lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE:
      if (!this->handle_deactivate()) {
        RCLCPP_DEBUG(get_logger(), "Shutdown failed during intermediate deactivation!");
        break;
      }
      [[fallthrough]];
    case lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE:
      if (!this->handle_cleanup()) {
        RCLCPP_DEBUG(get_logger(), "Shutdown failed during intermediate cleanup!");
        break;
      }
      [[fallthrough]];
    case lifecycle_msgs::msg::State::PRIMARY_STATE_UNCONFIGURED:
      if (!this->handle_shutdown()) {
        break;
      }
      //  TODO: reset and finalize all needed properties
      //  this->handlers_.clear();
      //  this->daemons_.clear();
      //  this->parameters_.clear();
      //  this->shutdown_ = true;
      return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
    default:
      RCLCPP_WARN(get_logger(), "Invalid transition 'shutdown' from state %s.", previous_state.label().c_str());
      break;
  }
  RCLCPP_ERROR(get_logger(), "Entering into the error processing transition state.");
  return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::ERROR;
}

bool LifecycleComponent::handle_shutdown() {
  bool result = this->on_shutdown_callback();
  return result && this->clear_signals();
}

bool LifecycleComponent::on_shutdown_callback() {
  return true;
}

rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
LifecycleComponent::on_error(const rclcpp_lifecycle::State& previous_state) {
  RCLCPP_DEBUG(this->get_logger(), "on_error called from previous state %s", previous_state.label().c_str());
  this->set_predicate("in_error_state", true);
  bool error_handled;
  try {
    error_handled = this->handle_error();
  } catch (const std::exception& ex) {
    RCLCPP_DEBUG(this->get_logger(), "Exception caught during on_error handling: %s", ex.what());
    error_handled = false;
  }
  if (!error_handled) {
    RCLCPP_ERROR(get_logger(), "Error processing failed! Entering into the finalized state.");
    // TODO: reset and finalize all needed properties
    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::ERROR;
  }
  this->set_predicate("in_error_state", false);
  return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
}

bool LifecycleComponent::handle_error() {
  return this->on_error_callback();
}

bool LifecycleComponent::on_error_callback() {
  return true;
}

bool LifecycleComponent::configure_outputs() {
  bool success = true;
  for (auto& [name, interface]: this->outputs_) {
    try {
      auto topic_name = this->get_parameter_value<std::string>(name + "_topic");
      RCLCPP_DEBUG_STREAM(this->get_logger(),
                          "Configuring output '" << name << "' with topic name '" << topic_name << "'.");
      auto message_pair = interface->get_message_pair();
      switch (message_pair->get_type()) {
        case MessageType::BOOL: {
          auto publisher = this->create_publisher<std_msgs::msg::Bool>(topic_name, this->qos_);
          interface = std::make_shared<PublisherHandler<rclcpp_lifecycle::LifecyclePublisher<std_msgs::msg::Bool>,
                                                        std_msgs::msg::Bool>>(
              PublisherType::LIFECYCLE_PUBLISHER, publisher
          )->create_publisher_interface(message_pair);
          break;
        }
        case MessageType::FLOAT64: {
          auto publisher = this->create_publisher<std_msgs::msg::Float64>(topic_name, this->qos_);
          interface = std::make_shared<PublisherHandler<rclcpp_lifecycle::LifecyclePublisher<std_msgs::msg::Float64>,
                                                        std_msgs::msg::Float64>>(
              PublisherType::LIFECYCLE_PUBLISHER, publisher
          )->create_publisher_interface(message_pair);
          break;
        }
        case MessageType::FLOAT64_MULTI_ARRAY: {
          auto publisher = this->create_publisher<std_msgs::msg::Float64MultiArray>(
              topic_name, this->qos_
          );
          interface =
              std::make_shared<PublisherHandler<rclcpp_lifecycle::LifecyclePublisher<std_msgs::msg::Float64MultiArray>,
                                                std_msgs::msg::Float64MultiArray>>(
                  PublisherType::LIFECYCLE_PUBLISHER, publisher
              )->create_publisher_interface(message_pair);
          break;
        }
        case MessageType::INT32: {
          auto publisher = this->create_publisher<std_msgs::msg::Int32>(topic_name, this->qos_);
          interface = std::make_shared<PublisherHandler<rclcpp_lifecycle::LifecyclePublisher<std_msgs::msg::Int32>,
                                                        std_msgs::msg::Int32>>(
              PublisherType::LIFECYCLE_PUBLISHER, publisher
          )->create_publisher_interface(message_pair);
          break;
        }
        case MessageType::STRING: {
          auto publisher = this->create_publisher<std_msgs::msg::String>(topic_name, this->qos_);
          interface = std::make_shared<PublisherHandler<rclcpp_lifecycle::LifecyclePublisher<std_msgs::msg::String>,
                                                        std_msgs::msg::String>>(
              PublisherType::LIFECYCLE_PUBLISHER, publisher
          )->create_publisher_interface(message_pair);
          break;
        }
        case MessageType::ENCODED_STATE: {
          auto publisher = this->create_publisher<modulo_core::EncodedState>(topic_name, this->qos_);
          interface = std::make_shared<PublisherHandler<rclcpp_lifecycle::LifecyclePublisher<modulo_core::EncodedState>,
                                                        modulo_core::EncodedState>>(
              PublisherType::LIFECYCLE_PUBLISHER, publisher
          )->create_publisher_interface(message_pair);
          break;
        }
      }
    } catch (const modulo_core::exceptions::CoreException& ex) {
      success = false;
      RCLCPP_ERROR_STREAM(this->get_logger(), "Failed to configure output '" << name << "': " << ex.what());
    }
  }
  return success;
}

bool LifecycleComponent::clear_signals() {
  RCLCPP_DEBUG(this->get_logger(), "Clearing all inputs and outputs.");
  this->inputs_.clear();
  this->outputs_.clear();
  return true;
}

bool LifecycleComponent::activate_outputs() {
  bool success = true;
  for (auto const& [name, interface]: this->outputs_) {
    try {
      interface->activate();
    } catch (const modulo_core::exceptions::CoreException& ex) {
      success = false;
      RCLCPP_ERROR_STREAM(this->get_logger(), "Failed to activate output '" << name << "': " << ex.what());
    }
  }
  RCLCPP_DEBUG(this->get_logger(), "All outputs activated.");
  return success;
}

bool LifecycleComponent::deactivate_outputs() {
  bool success = true;
  for (auto const& [name, interface]: this->outputs_) {
    try {
      interface->deactivate();
    } catch (const modulo_core::exceptions::CoreException& ex) {
      success = false;
      RCLCPP_ERROR_STREAM(this->get_logger(), "Failed to deactivate output '" << name << "': " << ex.what());
    }
  }
  RCLCPP_DEBUG(this->get_logger(), "All outputs deactivated.");
  return success;
}
}// namespace modulo_components

File: source/modulo_components/modulo_components/lifecycle_component.py
from typing import TypeVar

import clproto
from lifecycle_msgs.msg import State
from modulo_components.component_interface import ComponentInterface
from modulo_components.exceptions import AddSignalError
from rclpy.lifecycle import LifecycleNodeMixin, LifecycleState
from rclpy.lifecycle.node import TransitionCallbackReturn

MsgT = TypeVar('MsgT')


class LifecycleComponent(ComponentInterface, LifecycleNodeMixin):
    """
    Class to represent a LifecycleComponent in python, following the same logic pattern
    as the C++ modulo_components::LifecycleComponent class.
    """

    def __init__(self, node_name: str, enable_communication_interface=True, *args, **kwargs):
        """
        Constructs all the necessary attributes and declare all the parameters.

        :param node_name: The name of the node to be passed to the base Node class
        """
        ComponentInterface.__init__(self, node_name, *args, **kwargs)
        LifecycleNodeMixin.__init__(self, enable_communication_interface=enable_communication_interface, *args,
                                    **kwargs)

        self.add_predicate("is_unconfigured", lambda: self.get_state().state_id == State.PRIMARY_STATE_UNCONFIGURED)
        self.add_predicate("is_inactive", lambda: self.get_state().state_id == State.PRIMARY_STATE_INACTIVE)
        self.add_predicate("is_active", lambda: self.get_state().state_id == State.PRIMARY_STATE_ACTIVE)
        self.add_predicate("is_finalized", lambda: self.get_state().state_id == State.PRIMARY_STATE_FINALIZED)

    def get_state(self) -> LifecycleState:
        """
        Get the current state of the component.

        :return: The current state
        """
        return LifecycleState(self._state_machine.current_state[1], self._state_machine.current_state[0])

    def on_configure(self, previous_state: LifecycleState) -> TransitionCallbackReturn:
        """
        Transition callback for state 'Configuring'.

        on_configure callback is called when the lifecycle component enters the 'Configuring' transition state.
        The component must be in the 'Unconfigured' state.
        Depending on the return value of this function, the component may either transition to the 'Inactive' state
        via the 'configure' transition, stay 'Unconfigured' or go to 'ErrorProcessing'.
        TRANSITION_CALLBACK_SUCCESS transitions to 'Inactive'
        TRANSITION_CALLBACK_FAILURE transitions to 'Unconfigured'
        TRANSITION_CALLBACK_ERROR or any uncaught exceptions to 'ErrorProcessing'
        """
        self.get_logger().debug(f"on_configure called from previous state {previous_state.label}.")
        if previous_state.state_id != State.PRIMARY_STATE_UNCONFIGURED:
            self.get_logger().warn(f"Invalid transition 'configure' from state {previous_state.label}")
            return TransitionCallbackReturn.FAILURE
        if not self.__handle_configure():
            self.get_logger().warn("Configuration failed! Reverting to the unconfigured state.")
            if self.__handle_cleanup():
                return TransitionCallbackReturn.FAILURE
            else:
                self.get_logger().error(
                    "Could not revert to the unconfigured state! Entering into the error processing transition state.")
                return TransitionCallbackReturn.ERROR
        return TransitionCallbackReturn.SUCCESS

    def __handle_configure(self) -> bool:
        """
        Handle the configure transition by related transition steps and invoking the user callback.

        :return: True if configuration is successful, false otherwise
        """
        return self.on_configure_callback() and self.__configure_outputs()

    def on_configure_callback(self) -> bool:
        """
        Steps to execute when configuring the component. This method can be overridden by derived Component classes.
        Configuration generally involves reading parameters and adding inputs and outputs.

        :return: True if configuration is successful, false otherwise
        """
        return True

    def on_cleanup(self, previous_state: LifecycleState) -> TransitionCallbackReturn:
        """
        Transition callback for state 'CleaningUp'.

        on_cleanup callback is called when the lifecycle component enters the 'CleaningUp' transition state.
        The component must be in the 'Inactive' state.
        Depending on the return value of this function, the component may either transition to the 'Unconfigured' state
        via the 'cleanup' transition or go to 'ErrorProcessing'.
        TRANSITION_CALLBACK_SUCCESS transitions to 'Unconfigured'
        TRANSITION_CALLBACK_FAILURE, TRANSITION_CALLBACK_ERROR or any uncaught exceptions to 'ErrorProcessing'
        """
        self.get_logger().debug(f"on_cleanup called from previous state {previous_state.label}.")
        if previous_state.state_id != State.PRIMARY_STATE_INACTIVE:
            self.get_logger().warn(f"Invalid transition 'cleanup' from state {previous_state.label}")
        if not self.__handle_cleanup():
            self.get_logger().warn("Cleanup failed! Entering into the error processing transition state.")
            return TransitionCallbackReturn.ERROR
        return TransitionCallbackReturn.SUCCESS

    def __handle_cleanup(self) -> bool:
        """
        Handle the cleanup transition by related transition steps and invoking the user callback.

        :return: True if cleanup is successful, false otherwise
        """
        return self.on_cleanup_callback()

    def on_cleanup_callback(self) -> bool:
        """
        Steps to execute when cleaning up the component. This method can be overridden by derived Component classes.
        Cleanup generally involves resetting the properties and states to initial conditions.

        :return: True if cleanup is successful, false otherwise
        """
        return True

    def on_activate(self, previous_state: LifecycleState) -> TransitionCallbackReturn:
        """
        Transition callback for state 'Activating'.

        on_activate callback is called when the lifecycle component enters the 'Activating' transition state.
        The component must be in the 'Inactive' state.
        Depending on the return value of this function, the component may either transition to the 'Active' state
        via the 'activate' transition, stay 'Inactive' or go to 'ErrorProcessing'.
        TRANSITION_CALLBACK_SUCCESS transitions to 'Active'
        TRANSITION_CALLBACK_FAILURE transitions to 'Inactive'
        TRANSITION_CALLBACK_ERROR or any uncaught exceptions to 'ErrorProcessing'
        """
        self.get_logger().debug(f"on_activate called from previous state {previous_state.label}.")
        if previous_state.state_id != State.PRIMARY_STATE_INACTIVE:
            self.get_logger().warn(f"Invalid transition 'activate' from state {previous_state.label}")
            return TransitionCallbackReturn.FAILURE
        if not self.__handle_activate():
            self.get_logger().warn("Activation failed! Reverting to the inactive state.")
            # perform deactivation actions to ensure the component is inactive
            if self.__handle_deactivate():
                return TransitionCallbackReturn.FAILURE
            else:
                self.get_logger().error(
                    "Could not revert to the inactive state! Entering into the error processing transition state.")
                return TransitionCallbackReturn.ERROR
        return TransitionCallbackReturn.SUCCESS

    def __handle_activate(self) -> bool:
        """
        Handle the activate transition by related transition steps and invoking the user callback.

        :return: True if activation is successful, false otherwise
        """
        return self.on_activate_callback()

    def on_activate_callback(self) -> bool:
        """
        Steps to execute when activating the component. This method can be overridden by derived Component classes.
        Activation generally involves final setup steps before the on_step callback is periodically evaluated.

        :return: True if activation is successful, false otherwise
        """
        return True

    def on_deactivate(self, previous_state: LifecycleState) -> TransitionCallbackReturn:
        """
        Transition callback for state 'Deactivating'.

        on_deactivate callback is called when the lifecycle component enters the 'Deactivating' transition state.
        The component must be in the 'Active' state.
        Depending on the return value of this function, the component may either transition to the 'Inactive' state
        via the 'deactivate' transition or go to 'ErrorProcessing'.
        TRANSITION_CALLBACK_SUCCESS transitions to 'Inactive'
        TRANSITION_CALLBACK_FAILURE, TRANSITION_CALLBACK_ERROR or any uncaught exceptions to 'ErrorProcessing'
        """
        self.get_logger().debug(f"on_deactivate called from previous state {previous_state.label}.")
        if previous_state.state_id != State.PRIMARY_STATE_ACTIVE:
            self.get_logger().warn(f"Invalid transition 'deactivate' from state {previous_state.label}")
            return TransitionCallbackReturn.FAILURE
        if not self.__handle_deactivate():
            self.get_logger().warn("Deactivation failed! Reverting to the inactive state.")
            return TransitionCallbackReturn.ERROR
        return TransitionCallbackReturn.SUCCESS

    def __handle_deactivate(self) -> bool:
        """
        Handle the deactivate transition by related transition steps and invoking the user callback.

        :return: True if deactivation is successful, false otherwise
        """
        return self.on_deactivate_callback()

    def on_deactivate_callback(self) -> bool:
        """
        Steps to execute when deactivating the component. This method can be overridden by derived Component classes.
        Deactivation generally involves any steps to reset the component to an inactive state.

        :return: True if deactivation is successful, false otherwise
        """
        return True

    def on_shutdown(self, previous_state: LifecycleState) -> TransitionCallbackReturn:
        """
        Transition callback for state 'ShuttingDown'.

        on_shutdown callback is called when the lifecycle component enters the 'ShuttingDown' transition state.
        The component must be in the 'Unconfigured', 'Inactive' and 'Active' states.
        Depending on the return value of this function, the component may either transition to the 'Finalized' state
        via the 'shutdown' transition or go to 'ErrorProcessing'.
        TRANSITION_CALLBACK_SUCCESS transitions to 'Finalized'
        TRANSITION_CALLBACK_FAILURE, TRANSITION_CALLBACK_ERROR or any uncaught exceptions to 'ErrorProcessing'
        """
        self.get_logger().debug(f"on_shutdown called from previous state {previous_state.label}.")
        if previous_state.state_id == State.PRIMARY_STATE_FINALIZED:
            return TransitionCallbackReturn.SUCCESS
        if previous_state.state_id == State.PRIMARY_STATE_ACTIVE:
            if not self.__handle_deactivate():
                self.get_logger().debug("Shutdown failed during intermediate deactivation!")
        if previous_state.state_id == State.PRIMARY_STATE_INACTIVE:
            if not self.__handle_cleanup():
                self.get_logger().debug("Shutdown failed during intermediate cleanup!")
        if previous_state.state_id == State.PRIMARY_STATE_UNCONFIGURED:
            if not self.__handle_shutdown():
                self.get_logger().error("Entering into the error processing transition state.")
                return TransitionCallbackReturn.ERROR
            # TODO reset and finalize all properties
            return TransitionCallbackReturn.SUCCESS
        self.get_logger().warn(f"Invalid transition 'shutdown' from state {previous_state.label}.")
        return TransitionCallbackReturn.ERROR

    def __handle_shutdown(self) -> bool:
        """
        Handle the shutdown transition by related transition steps and invoking the user callback.

        :return: True if shutdown is successful, false otherwise
        """
        return self.on_shutdown_callback()

    def on_shutdown_callback(self) -> bool:
        """
        Steps to execute when shutting down the component. This method can be overridden by derived Component classes.
        Shutdown generally involves the destruction of any threads or properties not handled by the base class.

        :return: True if shutdown is successful, false otherwise
        """
        return True

    def on_error(self, previous_state: LifecycleState) -> TransitionCallbackReturn:
        """
        Transition callback for state 'ErrorProcessing'.

        on_shutdown callback is called when the lifecycle component enters the 'ErrorProcessing' transition state.
        This transition can originate from any step.
        Depending on the return value of this function, the component may either transition to the 'Unconfigured' state
        or go to 'Finalized'.
        TRANSITION_CALLBACK_SUCCESS transitions to 'Unconfigured'
        TRANSITION_CALLBACK_FAILURE transitions to 'Finalized'
        TRANSITION_CALLBACK_ERROR should not be returned, and any exceptions should be caught and returned as a failure
        """
        self.get_logger().debug(f"on_error called from previous state {previous_state.label}.")
        self.set_predicate("in_error_state", True)
        error_handled = False
        try:
            error_handled = self.__handle_error()
        except Exception as e:
            self.get_logger().debug(f"Exception caught during on_error handling: {e}")
            error_handled = False
        if not error_handled:
            self.get_logger().error("Error processing failed! Entering into the finalized state.")
            # TODO reset and finalize all needed properties
            return TransitionCallbackReturn.ERROR
        self.set_predicate("in_error_state", False)
        return TransitionCallbackReturn.SUCCESS

    def __handle_error(self) -> bool:
        """
        Handle the error transition by related transition steps and invoking the user callback.

        :return: True if error handling is successful, false otherwise
        """
        return self.on_error_callback()

    def on_error_callback(self) -> bool:
        """
        Steps to execute when handling errors. This method can be overridden by derived Component classes.
        Error handling generally involves recovering and resetting the component to an unconfigured state.

        :return: True if error handling is successful, false otherwise
        """
        return True

    def _step(self):
        """
        Step function that is called periodically and publishes predicates, outputs, evaluates daemon callbacks, and
        calls the on_step function.
        """
        try:
            self._publish_predicates()
            if self.get_state().state_id == State.PRIMARY_STATE_ACTIVE:
                self._publish_outputs()
                self._evaluate_periodic_callbacks()
                self.on_step_callback()
        except Exception as e:
            self.get_logger().error(f"Failed to execute step function: {e}", throttle_duration_sec=1.0)
            # TODO handle error in lifecycle component

    def on_step_callback(self):
        """
        Steps to execute periodically. To be redefined by derived classes.
        """
        pass

    def __configure_outputs(self) -> bool:
        """
        Configure all outputs.

        :return: True if configuration was successful
        """
        success = True
        for signal_name, output_dict in self._outputs.items():
            try:
                topic_name = self.get_parameter_value(signal_name + "_topic")
                self.get_logger().error(f"Configuring output '{signal_name}' with topic name '{topic_name}'.")
                publisher = self.create_publisher(output_dict["message_type"], topic_name, self._qos)
                self._outputs[signal_name]["publisher"] = publisher
            except Exception as e:
                success = False
                self.get_logger().debug(f"Failed to configure output '{signal_name}': {e}")
        return success

    def add_output(self, signal_name: str, data: str, message_type: MsgT,
                   clproto_message_type=clproto.MessageType.UNKNOWN_MESSAGE, default_topic="", fixed_topic=False):
        """
        Add an output signal of the component.

        :param signal_name: Name of the output signal
        :param data: Name of the attribute to transmit over the channel
        :param message_type: The ROS message type of the output
        :param clproto_message_type: The clproto message type, if applicable
        :param default_topic: If set, the default value for the topic name to use
        :param fixed_topic: If true, the topic name of the output signal is fixed
        """
        if self.get_state().state_id not in [State.PRIMARY_STATE_UNCONFIGURED, State.PRIMARY_STATE_INACTIVE]:
            self.get_logger().warn(f"Adding output in state {self.get_state().label} is not allowed.",
                                   throttle_duration_sec=1.0)
            return
        try:
            parsed_signal_name = self._create_output(signal_name, data, message_type, clproto_message_type,
                                                     default_topic, fixed_topic)
            topic_name = self.get_parameter_value(parsed_signal_name + "_topic")
            self.get_logger().debug(f"Adding output '{parsed_signal_name}' with topic name '{topic_name}'.")
        except AddSignalError as e:
            self.get_logger().error(f"Failed to add output '{signal_name}': {e}")
